### PR TITLE
GEOMESA-1739 Hbase Server Side Filter Implementation

### PR DESCRIPTION
**Pull request by spandanagrawal**
This commit implements HBase Server Side Filtering. Server side filtering is on by default, but is functioning correctly and can be turned on.

All requested changes in the old PR have been made! The commits have been rebased on locationtech/master to remove merge conflicts with the base branch. All tests have been run and passed.

Transforms will be added as a separate PR later. This pull request only implements HBase Filters.

If there are any other changes required please let me know!

### CODE REVIEW

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala [hunk 1]
```diff
@@ -55,12 +56,16 @@ case class EmptyPlan(filter: HBaseFilterStrategyType) extends HBaseQueryPlan {
...
+case class ScanPlan(sft: SimpleFeatureType,
+                    filter: HBaseFilterStrategyType,
                     table: TableName,
                     ranges: Seq[Scan],
                     remoteFilters: Seq[HBaseFilter] = Nil,
                     resultsToFeatures: Iterator[Result] => Iterator[SimpleFeature]) extends HBaseQueryPlan {
   override def scan(ds: HBaseDataStore): CloseableIterator[SimpleFeature] = {
+    val remoteFilters = filter.filter.map { filter =>
+      new JSimpleFeatureFilter(sft, filter)
+    }.toSeq
```
Comment (jnh5y) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: Ok, something is at odds here.

There should either be remoteFilters computed here or passed in.  Which way are we going with that?

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/utils/HBaseBatchScan.scala [hunk 2]
```diff
@@ -15,7 +15,12 @@ import org.apache.hadoop.hbase.client._
 import org.apache.hadoop.hbase.filter.Filter
 import org.locationtech.geomesa.index.utils.AbstractBatchScan
 
-class HBaseBatchScan(connection: Connection, tableName: TableName, ranges: Seq[Scan], threads: Int, buffer: Int, remoteFilters: Seq[Filter] = Nil)
+class HBaseBatchScan(connection: Connection,
```
Comment (jnh5y) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/utils/HBaseBatchScan.scala]: Is this change needed or just style?  If the latter, remove it.

Comment (spandanagrawal) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/utils/HBaseBatchScan.scala]: Yes it was just style, will remove it

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala (hunk 1 above)
Comment (spandanagrawal) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: Fixed in new commit

File: geomesa-hbase/geomesa-hbase-datastore/src/main/java/org/locationtech/geomesa/hbase/filters/JSimpleFeatureFilter.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+import org.geotools.filter.text.ecql.ECQL;
+import org.locationtech.geomesa.features.interop.SerializationOptions;
+import org.locationtech.geomesa.features.kryo.KryoFeatureSerializer;
+import org.locationtech.geomesa.utils.geotools.SimpleFeatureTypes;
+import org.opengis.feature.simple.SimpleFeature;
+import org.opengis.feature.simple.SimpleFeatureType;
+
+import java.io.IOException;
+
+public class JSimpleFeatureFilter extends FilterBase {
```
Comment (jahhulbert-ccri) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/java/org/locationtech/geomesa/hbase/filters/JSimpleFeatureFilter.java]: Prob still would remove ```J``` prefix

Comment (jnh5y) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/java/org/locationtech/geomesa/hbase/filters/JSimpleFeatureFilter.java]: @jahhulbert-ccri eh, we could change the name in the next PR when Transforms come in.  (That PR would need to change the class name since there'd be support for transforms...)

File: geomesa-hbase/geomesa-hbase-datastore/src/main/java/org/locationtech/geomesa/hbase/filters/JSimpleFeatureFilter.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    private org.opengis.filter.Filter filter;
+    private String filterString;
+
+    public JSimpleFeatureFilter(String sftString, String filterString) {
+        this.sftString = sftString;
+        sft = SimpleFeatureTypes.createType("", sftString);
+        serializer = new KryoFeatureSerializer(sft, SerializationOptions.withoutId());
+
+        this.filterString = filterString;
+        if (filterString != null && filterString != "") {
```
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/java/org/locationtech/geomesa/hbase/filters/JSimpleFeatureFilter.java]: potential bug here if filterString is empty, filter won't be set but null check on line 57 will still pass, causing NPE on line 61

File: geomesa-hbase/geomesa-hbase-datastore/src/main/java/org/locationtech/geomesa/hbase/filters/JSimpleFeatureFilter.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+    @Override
+    public Cell transformCell(Cell v) throws IOException {
+        return super.transformCell(v);
+    }
+
+    // TODO: Add static method to compute byte array from SFT and Filter.
+    @Override
+    public byte[] toByteArray() throws IOException {
+      return Bytes.add(getLengthArray(sftString), getLengthArray(filterString));
```
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/java/org/locationtech/geomesa/hbase/filters/JSimpleFeatureFilter.java]: [minor] indentation is off

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala [hunk 6]
```diff
@@ -55,7 +55,8 @@ case class EmptyPlan(filter: HBaseFilterStrategyType) extends HBaseQueryPlan {
   override def scan(ds: HBaseDataStore): CloseableIterator[SimpleFeature] = CloseableIterator.empty
 }
 
-case class ScanPlan(filter: HBaseFilterStrategyType,
+case class ScanPlan(sft: SimpleFeatureType,
```
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: i think i would revert this change unless you need it for something - it's missing from GetPlan, so if we actually need it we would have to add it there

File: geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/KryoLazyFilterTransformIterator.scala [hunk 7]
```diff
@@ -18,7 +18,6 @@ import org.geotools.factory.Hints
 import org.geotools.filter.text.ecql.ECQL
 import org.locationtech.geomesa.accumulo.AccumuloFeatureIndexType
 import org.locationtech.geomesa.accumulo.index.AccumuloFeatureIndex
-import org.locationtech.geomesa.features.SerializationOption.SerializationOptions
```
Comment (elahrvivaz) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/KryoLazyFilterTransformIterator.scala]: oops, need to revert this change it's making the build fail

File: geomesa-hbase/geomesa-hbase-datastore/src/main/java/org/locationtech/geomesa/hbase/filters/JSimpleFeatureFilter.java (hunk 4 above)
Comment (spandanagrawal) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/java/org/locationtech/geomesa/hbase/filters/JSimpleFeatureFilter.java]: If the filterString is empty, filter won't be set on line 41 and will remain null. Then on line 57 the null check will pass and it will simply include that key value pair. It shouldn't reach line 61.

Is there something wrong with my understanding?

File: geomesa-hbase/geomesa-hbase-datastore/src/main/java/org/locationtech/geomesa/hbase/filters/JSimpleFeatureFilter.java (hunk 5 above)
Comment (spandanagrawal) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/java/org/locationtech/geomesa/hbase/filters/JSimpleFeatureFilter.java]: Fixed

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala (hunk 6 above)
Comment (spandanagrawal) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala]: Fixed

File: geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/KryoLazyFilterTransformIterator.scala (hunk 7 above)
Comment (spandanagrawal) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/iterators/KryoLazyFilterTransformIterator.scala]: Fixed

File: geomesa-hbase/geomesa-hbase-datastore/src/main/java/org/locationtech/geomesa/hbase/filters/JSimpleFeatureFilter.java (hunk 4 above)
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/java/org/locationtech/geomesa/hbase/filters/JSimpleFeatureFilter.java]: oops, you're right, sorry about that

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala [hunk 8]
```diff
@@ -176,8 +176,10 @@ trait HBaseFeatureIndex extends HBaseFeatureIndexType
     import org.locationtech.geomesa.index.conf.QueryHints.RichHints
 
     /** This function is used to implement custom client filters for HBase **/
-      val toFeatures = resultsToFeatures(sft, ecql, hints.getTransform)
-      ScanConfig(Nil, toFeatures)
-
+      val toFeatures = resultsToFeatures(sft, None, hints.getTransform)
+      val remoteFilters = filter.filter.map { filter =>
```
Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala]: This should use `ecql` instead of `filter.filter` - they may not be the same, `ecql` is the remaining predicate that needs to be applied, which `filter.filter` is the entire predicate, some of which may already have been handled with range planning.

**Comment by elahrvivaz** [comment]
looks good, noticed one more thing then I'll merge

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala (hunk 8 above)
Comment (spandanagrawal) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala]: So that should be changed to `ecql.map` instead of `filter.filter.map`?

Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala]: yep, exactly

Comment (spandanagrawal) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala]: Okay will fix it and push another commit by today!

**Comment by spandanagrawal** [comment]
All final changes have been made!

File: geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala (hunk 8 above)
Comment (MutahirKazmi) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala]: @elahrvivaz will this change not always implement JSimpleFeatureFilter? Shouldn't there be an option to run the default version and based on some hint, JSimpleFeatureFilter?

Comment (elahrvivaz) [inline geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/index/HBaseFeatureIndex.scala]: @MutahirKazmi yes good point. @jnh5y thoughts? My understanding was we wanted to merge this one first, then merge the PR that toggles the remote filter. That will put us in a temporary head state where filters will always be run.
